rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,318 @@
+
+**Note:** Clustering and updates conflict resolution is not in the scope of non-blocking concurrency control, you still need the OCC for course-grained failures and retries. 
+
+## Appendix
+
+### Pseudocode for New File Slicing
+
+Assume we have the following inputs to build the file slices:
+1. `base_files_by_file_id`: a map of filegroup id to list of base files.
+2. `log_files_by_file_id`: a map of feilgroup id to list of log files.
+3. `timeline`: Hudi write timeline .
+
+The pseudocode below introduces a notion of **file slice barriers**, which contain a list of instant (start) times of 
+the base files in descending order. Barriers will help in demarcating file slices. The below pseudocode builds the file 
+slices per file group. Building file groups from file slices is not shown but can be easily done.
+
+```python
+# new file slicing
+def build_file_slices(base_files_by_file_id, log_files_by_file_id, timeline):
+    # get set of all filegroup ids
+    file_id_set = base_files_by_file_id.keys
+    file_id_set.add_all(log_files_by_file_id.keys)
+    
+    for file_id in file_id_set:
+        # sort the base files by descending order of instant (start) time, i.e. last written base file first
+        base_files_in_file_id = base_files_by_file_id[file_id].sort(BASE_FILE_REVERSE_COMPARATOR)
+        # filter out log files that have been log-compacted
+        log_files_in_file_id = handle_log_compaction(log_files_by_file_id[file_id], timeline)
+        # sort the log files by ascending order of completion time
+        log_files_in_file_id.sort(LOG_FILE_COMPARATOR)
+        # get list of file slice barriers for this fielgroup id
+        file_slice_barriers = get_file_slice_barriers(base_files_in_file_id, log_files_in_file_id)    
+        # build file slices
+        file_slices = []
+        for log_file in log_files_in_file_id:
+            file_slice = find_file_slice(log_file, file_slice_barriers)
+            file_slices.add(file_slice)
+            
+            
+# Given all log files for a file id, filter out such log files that have been log-compacted.
+def handle_log_compaction(log_files_in_file_id, timeline):
+    log_compaction_instants = timeline.get_completed_log_compaction()
+    for log_compaction_instant in log_compaction_instant:
+        log_files_compacted = get_log_files_compacted(log_compaction_instant)
+        log_files_in_file_id.remove(log_files_compacted)
+        
+    return log_files_in_file_id
+
+
+# Given base files and log files for a filegroup id, return a list containing file slice barriers. 
+def get_file_slice_barriers(base_files_in_file_id, log_files_in_file_id):
+    file_slice_barriers = []
+    if base_files_in_file_id.size > 0:
+        for base_file in base_files_in_file_id:
+            file_slice_barriers.add(instant_time(base_file))
```

Comment:
this is so that we can associate any log file with completion time < instant_time of `base_file` with the prev file slice.